src/derivative_operators/multi_dim_bc_operators.jl

```diff
@@ -54,6 +54,44 @@ struct ComposedMultiDimBC{T, B<:AtomicBC{T}, N,M} <: MultiDimensionalBC{T, N}
     BCs::Vector{Array{B, M}}
 end
 
+struct PartiallyComposedMultiDimBC{T,B<:MultiDimDirectionalBC,N,L} <:MultiDimensionalBC{T,N}
```

Comment:
`+(a,b,c)` isn't equivalent to `a+b+c` is it? Else I think a partially composed BC is necessary so that julia can do `(a + b)::PartiallyComposedMultiDimBC + c::MultiDimBC` as an intermediate step between having a single dimension BC and a fully Compsed BC.

Another reason for the PartiallyComposedBPAs is due to the position in the` BCs` Array in a `ComposedMultiDimBC` being identified with a particular dimension to extend along, and a requirement at the moment that a BC is present for every dimension. This is also the case for the `CompsedBoundaryPaddedArray`.

We could get around this with a `MissingBC` type that acts as a placeholder within a `ComposedMultiDimBC`. This would extend an `Array` with  `MissingArray`s, which act as placeholders within `upper` and `lower` in a `ComposedBoundaryPaddedArray`.
I think that this would lead to ambiguous types in the `ComposedBoundaryPaddedArray` though, and would make it more complex to write code that works on them.

Reply:
Then I'll replace all of that with  `Base.:+(BCs::MultiDimDirectionalBCs...) =  compose(BCs...)`. This is fine, as long as we are understanding that `(Qx+Qz)` wouldn't work when the u to be extended is 3D, so in that case doing something like `Lx*Qx + Lz*Qz` would be necessary.

Reply:
> This is fine, as long as we are understanding that (Qx+Qz) wouldn't work when the u to be extended is 3D

Why wouldn't it work? It would catch finitely many multiplies.

Reply:
It is just because of the way that `ComposedMultiDimBC` is set up at the moment, also `ComposedBoundaryPaddedArray` since these currently rely on there being an extension for every dimension.

It would be possible to allow the Composed types it to support any number of extended and unextended dimensions, at the cost of quite a bit of additional complexity for any algorithms that need to index this type. This would require a significant rewrite.

This complexity mostly comes in the form of checks to see whether to index in to `A.upper[2][i,j]` or `A.u[i, end, j]` when doing a boundary convolution for example. 

The question is whether this use case is common enough to justify throwing out specialised methods for fully composed BC operators and PaddedArrays

Reply:
alright, then a `composed` function makes sense.

Reply:
I will split out the `PartiallyComposed` BCs in to another PR then - we can discuss further whether or not having this kind of functionality is a priority.